internal/pkg/runtime/engine/singularity/container_linux.go

```diff
@@ -1248,8 +1248,27 @@ func (c *container) addBindsMount(system *mount.System) error {
 	flags := uintptr(syscall.MS_BIND | c.suidFlag | syscall.MS_NODEV | syscall.MS_REC)
 
 	if c.engine.EngineConfig.GetContain() {
-		sylog.Debugf("Skipping bind mounts as contain was requested")
-		return nil
+		const hostsPath = "/etc/hosts"
+		hosts := hostsPath
+
+		// handle special case for /etc/hosts as it is required,
+		// if no network namespace was requested we simply bind
+		// /etc/hosts from host, if network namespace is requested
+		// we create a minimal default hosts for localhost resolution
+		if !c.netNS {
+			sylog.Debugf("Binding /etc/hosts only with contain")
+		} else {
+			sylog.Debugf("Skipping bind mounts as contain was requested")
+
+			sylog.Verbosef("Binding staging /etc/hosts as contain is set")
+			if err := c.session.AddFile(hostsPath, files.DefaultHosts()); err != nil {
+				return fmt.Errorf("while adding /etc/hosts staging file: %s", err)
+			}
+			hosts, _ = c.session.GetPath(hostsPath)
+		}
+
+		system.Points.AddBind(mount.BindsTag, hosts, hostsPath, flags)
```

Comment:
In the loop below, you catch the error on `AddBind` is that not necessary here?

Reply:
Fixed